### ISSUE TITLE
Cannot transpile default angular SPA template in docker multistage build file

### ISSUE THREAD

**Issue by temporafugiunt** [dotnet/dotnet-docker#606]
## Steps to reproduce the issue
I outlined a complete recreation scenario in this github repository for this issue, the readme file outlines all the steps to recreate the issue, plus the repository already contains the source to recreate the issue.

https://github.com/temporafugiunt/AngularSPATemplateFailure

What it boils down to is whether I use *microsoft/dotnet:2.1-sdk* and install my own version of node or I use *microsoft/aspnetcore-build:2.0.8-2.1.200-nanoserver-sac2016* and use the older version of node when I create a brand new Angular SPA app using the latest .NET Core SPA templates when I create a multistage build file for that template it fails to build the angular portion during a *dotnet publish* operation and fails with the error given below.

## Expected behavior

*dotnet publish* operation should work exactly as it does when building on windows 2016 server with all the proper tools installed (2.1 SDK, node.js, VS Code, etc.) and should create the production ready website.

## Actual behavior
ng command fails to properly execute to create the output javascript and fails with this error:

```
  > failed_spa_build@0.0.0 build C:\app\failed-spa-build-src\ClientApp
  > ng build --extract-css "--prod"

  module.js:549
      throw err;
      ^

EXEC : error : Cannot find module 'C:\app\failed-spa-build-src\ClientApp\node_modules\@angular\cli\bin\ng' [C:\app\failed-spa-build-src\failed-spa-build.csproj]
      at Function.Module._resolveFilename (module.js:547:15)
      at Function.Module._load (module.js:474:25)
      at Function.Module.runMain (module.js:693:10)
      at startup (bootstrap_node.js:191:16)
      at bootstrap_node.js:612:3
  npm ERR! code ELIFECYCLE
  npm ERR! errno 1
  npm ERR! failed_spa_build@0.0.0 build: `ng build --extract-css "--prod"`
  npm ERR! Exit status 1
  npm ERR!
  npm ERR! Failed at the failed_spa_build@0.0.0 build script.
  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

  npm ERR! A complete log of this run can be found in:
  npm ERR!     C:\Users\ContainerAdministrator\AppData\Roaming\npm-cache\_logs\2018-06-25T20_23_36_785Z-debug.log
C:\app\failed-spa-build-src\failed-spa-build.csproj(43,5): error MSB3073: The command "npm run build -- --prod" exited with code 1.
The command 'powershell -Command $ErrorActionPreference = 'Stop'; $ProgressPreference = 'SilentlyContinue'; dotnet publish -c Release -o /app/failed-spa-build-pub' return
ed a non-zero code: 1
```

## Output of `docker version`

```
Client:
 Version:       17.06.2-ee-13
 API version:   1.30
 Go version:    go1.8.7
 Git commit:    ac44d73
 Built: Mon Jun  4 16:46:59 2018
 OS/Arch:       windows/amd64

Server:
 Engine:
  Version:      17.06.2-ee-13
  API version:  1.30 (minimum version 1.24)
  Go version:   go1.8.7
  Git commit:   ac44d73
  Built:        Mon Jun  4 16:58:47 2018
  OS/Arch:      windows/amd64
  Experimental: false
```

## Output of `docker info`

```
Containers: 46
 Running: 0
 Paused: 0
 Stopped: 46
Images: 177
Server Version: 17.06.2-ee-13
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.2312.amd64fre.rs1_release.180607-1919)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 5.999GiB
Name: jeby2016
ID: XY5B:2VRV:JPQ5:H7BZ:FUGR:COQO:TQIF:S6EG:BVQC:OFQE:EY6L:L46Z
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by temporafugiunt** [comment]
Doing some testing inside the container and I attempted to update the angular CLI version to 1.7.4 using this command:

```
npm install @angular/cli@1.7.4 --save-dev
```

and I recieved this failure which definitely seems relevant:

```
4578 verbose stack Error: Refusing to delete C:\app\failed-spa-build-src\ClientApp\node_modules\.bin\ng.cmd: ..\@angular\cli\bin\ng symlink target is not controlled by npm C:\app\failed-spa-build-src\ClientApp\node_modules\@angular\cli
4578 verbose stack     at clobberFail (C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\gentle-fs\lib\rm.js:121:12)
4578 verbose stack     at isSafeToRm (C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\gentle-fs\lib\rm.js:111:15)
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\gentle-fs\lib\rm.js:54:5
4578 verbose stack     at LOOP (C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\slide\lib\chain.js:7:26)
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\slide\lib\chain.js:18:7
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\gentle-fs\lib\rm.js:138:14
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\iferr\index.js:11:16
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\iferr\index.js:11:16
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\iferr\index.js:11:16
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\iferr\index.js:11:16
4578 verbose stack     at C:\build\node-v8.11.3-win-x64\node_modules\npm\node_modules\graceful-fs\polyfills.js:284:29
4578 verbose stack     at FSReqWrap.oncomplete (fs.js:152:21)
4579 verbose cwd C:\app\failed-spa-build-src\ClientApp
4580 verbose Windows_NT 10.0.14393
4581 verbose argv "C:\\build\\node-v8.11.3-win-x64\\node.exe" "C:\\build\\node-v8.11.3-win-x64\\node_modules\\npm\\bin\\npm-cli.js" "install" "@angular/cli@1.7.4" "--save-dev"
4582 verbose node v8.11.3
4583 verbose npm  v5.6.0
4584 error path C:\app\failed-spa-build-src\ClientApp\node_modules\.bin\ng.cmd
4585 error code EEXIST
4586 error Refusing to delete C:\app\failed-spa-build-src\ClientApp\node_modules\.bin\ng.cmd: ..\@angular\cli\bin\ng symlink target is not controlled by npm C:\app\failed-spa-build-src\ClientApp\node_modules\@angular\cli
4587 error File exists: C:\app\failed-spa-build-src\ClientApp\node_modules\.bin\ng.cmd
4588 error Move it away, and try again.
4589 verbose exit [ 1, true ]
```
Is this a symbolic link issue where it expects a link between files in **C:\app\failed-spa-build-src\ClientApp\node_modules\.bin** and **C:\app\failed-spa-build-src\ClientApp\node_modules\@angular\cli\bin\ng** but those were not created because of some restriction with the nanoserver container?

**Comment by MichaelSimons** [comment]
FWIW, I pulled your [source ](https://github.com/temporafugiunt/AngularSPATemplateFailure), changed the FROM in the Dockerfile to `microsoft/dotnet:2.1-sdk-nanoserver-sac2016` and was able to build the Dockerfile cleanly.  The reason I changed the base image is that I'm on a win10 machine and the 2.1-sdk resolves to an Windows Server 1803 image which doesn't have PowerShell.  

The one thing that comes to mind is that I didn't build the app locally like you mention in the instructions.  Is the `.dockerignore` file excluding everything that is necessary?  Can you try pulling your repo into a clean location and building to confirm?

**Comment by temporafugiunt** [comment]
WOW! Thank you MichaelSimons...

That was a total ID10T move on my part, I can't believe I did that. You are exactly right, the difference was that you didn't perform a local build but I did.

I copied that .dockerignore file from a sample project _**and didn't even think about it after that**_ (the sample project was NOT using a webpack or other javascript based build technology so the ignore was not complete).

****/node_modules/** was not defined in there as an ignored directory structure so it was pulling the files from the local file system which was creating conflicts with the build in the image (not totally sure why, I would think it would just skip the pull step). Once I added that and rebuilt with the **--no-cache** option it built.

So the issue was all my fault. Someone can go ahead and close this.
